pkg/api/types.go

```diff
@@ -1077,8 +1077,14 @@ type Event struct {
 	// Optional. The component reporting this event. Should be a short machine understandable string.
 	Source EventSource `json:"source,omitempty"`
 
-	// The time at which the client recorded the event. (Time of server receipt is in TypeMeta.)
-	Timestamp util.Time `json:"timestamp,omitempty"`
+	// The time at which the event was first recorded. (Time of server receipt is in TypeMeta.)
+	InitialTimestamp util.Time `json:"initialTimestamp,omitempty"`
+
+	// The time at which the most recent occurance of this event was recorded.
+	LastSeenTimestamp util.Time `json:"initialTimestamp,omitempty"`
+
+	// The number of times this event has occurred.
```

Comment:
See the detailed design proposal in #4073.

Short answer is compression will be best effort.

The compression is intended to happen on each kubelet instance. Each kubelet will maintain a hash of each unique event it sees (minus the transient info like timestamps/count) and if it sees the same event again it will update the previous instance in etcd instead of creating a new one.

That means that compression will not occur across kubelet restarts.

Also, as noted in the proposal, if, in the future, we decide to age out events from the kubelet events hash table, then events will only be compressed until they age out of the hash table, at which point any new instance of the event will create a new entry in etcd.